### PR TITLE
Implement ICS-26 routing in `{Validation, Execution}Context` for `ChanOpenTry`

### PR DESCRIPTION
**Pull request by plafer**
Work towards: #322

This implements ICS-26 routing in `{Validation, Execution}Context` for the `ChanOpenTry` message only. After this PR is merged, I'll implement the other messages in a similar manner.

Note that we prioritize here a *working* implementation. We will improve it once a working version is shipped (see #325).

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/core/context.rs [hunk 1]
```diff
@@ -126,7 +163,26 @@ mod val_exec_ctx {
...
                 }
                 .map_err(RouterError::ContextError),
-                MsgEnvelope::Channel(_message) => todo!(),
+                MsgEnvelope::Channel(message) => {
+                    let module_id = self.lookup_module(&message).map_err(ContextError::from)?;
+                    if !self.has_route(&module_id) {
+                        return Err(ChannelError::RouteNotFound)
+                            .map_err(ContextError::ChannelError)
+                            .map_err(RouterError::ContextError);
+                    }
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/context.rs:173]: By having this check here, We'll drop following lines in the future. Right? 
- https://github.com/cosmos/ibc-rs/blob/78297c39ac9a892ee6367dc2b96cdd881013a08a/crates/ibc/src/core/ics04_channel/handler.rs#L67-L77
- https://github.com/cosmos/ibc-rs/blob/aa97ced3f936dd0cb4bf586d340f36af4e5c986a/crates/ibc/src/core/ics04_channel/msgs.rs#L45-L58

File: crates/ibc/src/core/context.rs [hunk 2]
```diff
@@ -99,11 +105,42 @@ mod val_exec_ctx {
     };
     use crate::events::IbcEvent;
     use crate::timestamp::Timestamp;
-    use crate::{prelude::*, Height};
+    use crate::Height;
 
     use super::ContextError;
 
-    pub trait ValidationContext {
+    pub trait Router {
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/context.rs:113]: So, we'll finaly merge the `Router` trait under `ics26_routing` module with this new one? 
https://github.com/cosmos/ibc-rs/blob/380056be894a6edf8b0d9fc23175d92641407a16/crates/ibc/src/core/ics26_routing/context.rs#L248

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 3]
```diff
@@ -19,6 +19,8 @@ pub enum ChannelError {
     Connection(connection_error::ConnectionError),
     /// port error: `{0}`
     Port(port_error::PortError),
+    /// port `{port_id}` is unknown
+    PortNoSource { port_id: PortId },
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics04_channel/error.rs]: Couldn't we use the above `Port` variant instead? This looks the same as `UnknownPort`

File: crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs [hunk 4]
```diff
@@ -11,6 +11,112 @@ use crate::core::ics24_host::identifier::ChannelId;
...
+                    &msg.proof_chan_end_on_a,
+                    consensus_state_of_a_on_b.root(),
+                    port_id_on_a,
+                    &chan_id_on_a,
+                    &expected_chan_end_on_a,
+                )
+                .map_err(ChannelError::VerifyChannelFailed)?;
+        }
+
+        let chan_id_on_b = ChannelId::new(ctx_b.channel_counter()?);
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs]: IMO, returning `chan_id_on_b` here gives the false impression that it is a result or at least by-product of the validation process. The declaration can be made after the passed "validate" process, in "chan_open_try_validate". Therefore, `validate` would return Ok(())

File: crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs [hunk 5]
```diff
@@ -11,6 +11,112 @@ use crate::core::ics24_host::identifier::ChannelId;
...
+
+        Ok(chan_id_on_b)
+    }
+
+    pub fn execute<Ctx>(ctx_b: &mut Ctx) -> Result<ChannelId, ContextError>
+    where
+        Ctx: ExecutionContext,
+    {
+        let chan_id_on_b = ChannelId::new(ctx_b.channel_counter()?);
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs]: The same comment for `validate` applies here, so makes the `execute` no longer necessary in this case...

File: crates/ibc/src/core/context.rs [hunk 6]
```diff
@@ -126,7 +163,26 @@ mod val_exec_ctx {
...
+                    if !self.has_route(&module_id) {
+                        return Err(ChannelError::RouteNotFound)
+                            .map_err(ContextError::ChannelError)
+                            .map_err(RouterError::ContextError);
+                    }
+
+                    match message {
+                        ChannelMsg::OpenInit(_) => todo!(),
+                        ChannelMsg::OpenTry(message) => {
+                            chan_open_try_validate(self, module_id, message)
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/context.rs:178]: I thought how about calling `chan_open_try::validate` directly instead of adding another layer of calls by `chan_open_try_validate`? (or just keeping `chan_open_try_validate`) 
Can't we merge these two somehow? The reasoning behind is not clear, making the code a bit hard to read and understand.

**Review by Farhad-Shabani — COMMENTED** [review]
Thank you @plafer. 
I'll probably revisit this PR after getting your response so I can absorb the design. Hope it doesn't cause you any delays.

File: crates/ibc/src/core/context.rs (hunk 1 above)
Comment (plafer) [reply crates/ibc/src/core/context.rs:173]: yep!

File: crates/ibc/src/core/context.rs (hunk 2 above)
Comment (plafer) [reply crates/ibc/src/core/context.rs:113]: We will be removing the one in 'ics26_routing' completely

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 3 above)
Comment (plafer) [reply crates/ibc/src/core/ics04_channel/error.rs]: yes good point; I wanted to avoid using `PortError` because I think we should get rid of it, but this is better for now

[c7f10ad](https://github.com/cosmos/ibc-rs/pull/347/commits/c7f10adfa0024037669515b85f5936c4997fb425)

File: crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs (hunk 4 above)
Comment (plafer) [reply crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs]: Good point! The original intention was that making sure that the `channel_counter()` call succeeds is part of validation; however it's minor and I much prefer this too.

[6047f02](https://github.com/cosmos/ibc-rs/pull/347/commits/6047f02d2cce8b9661f0bcb85fd0dbc68928126e)

File: crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs (hunk 5 above)
Comment (plafer) [reply crates/ibc/src/core/ics04_channel/handler/chan_open_try.rs]: [28d04f0](https://github.com/cosmos/ibc-rs/pull/347/commits/28d04f0fe09cd12dc31897a2b4c340c80c0d90f9)

File: crates/ibc/src/core/context.rs (hunk 6 above)
Comment (plafer) [reply crates/ibc/src/core/context.rs:178]: The current design is based on the standard: `chan_open_try_validate()` is the validation part of the [ICS-26 router function](https://github.com/cosmos/ibc/blob/main/spec/core/ics-026-routing-module/README.md#channel-lifecycle-management) (which is currently incorrect in the spec), and `chan_open_try::validate()` is the validation part of the [core handler](https://github.com/cosmos/ibc/blob/main/spec/core/ics-004-channel-and-packet-semantics/README.md#opening-handshake).

In other words:
- `chan_open_try_validate()` calls both the core handler and the application/module callback (getting the channel id from the core handler,  and the channel version from the callback)
- `chan_open_try::validate()` is the core handler validation logic

Does this make sense? Do you still think it would be better to move away from this design? At the very least, using better naming to encode what I just explained would help.

Comment (Farhad-Shabani) [reply crates/ibc/src/core/context.rs:178]: On this basis, `chan_open_try_validate` operates more as a handler than a validator, shouldn't contain any validation step. Also, the latter link you sent seems to correspond more with module interfaces iiuc (e.g. `on_chan_open_try_validate`).
Btw, I don't think naming help with this issue. I propose all checkings be done in one place either in `chan_open_try::validate()` or `on_chan_open_try_validate`, like what we have now for `on_chan_open_try_execute` or like before structure for `validate()` `execute()`

Comment (Farhad-Shabani) [reply crates/ibc/src/core/context.rs:178]: Maybe not bad, keeping this change, and see what the implications will be when we refactor others. Beside this, it LGTM and your explanations are great, thanks 🙏

Comment (plafer) [reply crates/ibc/src/core/context.rs:178]: Indeed I think we should leave it as is for now because correctness doesn't seem to be a concern; we will focus on improving readability/maintainability once `Validation/ExecutionContext` have shipped. WDYT?

Comment (Farhad-Shabani) [reply crates/ibc/src/core/context.rs:178]: Agree, let's keep it for now 👍

**Review by Farhad-Shabani — APPROVED** [review]
👌
